src/emitter/batch_emitter.rs

```diff
@@ -0,0 +1,318 @@
+                            match Self::send_batch(batch, client).await {
+                                Ok(_) => println!("Batch sent"),
+                                Err(e) => println!("Error sending batch: {e}"),
+                            }
+                        }));
+                    }
+
+                    // On break, the emitter and runtime will be dropped
+                    //
+                    // Tokio will cancel any running tasks once the runtime is dropped, meaning any queued batches will be lost,
+                    // so we attempt to send any remaining batches before exiting
+                    EmitterMessage::Close => {
+                        for task in tokio_tasks {
+                            task.await.unwrap();
+                        }
+                        break;
+                    }
+                }
+
+                // Discard any completed tasks in the task list
+                tokio_tasks.retain(|t| !t.is_finished());
+            }
+        });
+    }
+}
+
+impl Drop for BatchEmitter {
+    fn drop(&mut self) {
+        // Get the join handle for the thread running the tokio runtime and wait for it to finish
+        //
+        // It's likely that the thread has already finished once the emitter loop has exited
+        if let Some(handle) = self.executor_handle.take() {
+            handle.join().unwrap();
+        }
+    }
+}
+
+impl Emitter for BatchEmitter {
+    /// Adds a payload to the event store
+    ///
+    /// This may also trigger sending a payload to the collector if the event store has enough events to fill a batch
+    fn add(&mut self, payload: PayloadBuilder) -> Result<(), Error> {
+        match self.event_store.lock() {
+            Ok(mut store) => match store.add(payload) {
+                Ok(_) => Ok(()),
+                Err(e) => Err(Error::EmitterError(e.to_string())),
+            },
+            Err(e) => Err(Error::EmitterError(e.to_string())),
+        }?;
+
+        let batch: Option<EventBatch> = {
+            let mut store_lock = self.event_store.lock().unwrap();
+            match store_lock.batch() {
+                Ok(batch) => Some(batch),
+                Err(_) => None,
+            }
+        };
+
+        if let Some(batch) = batch {
+            return match self.tx.try_send(EmitterMessage::Send(batch)) {
+                Ok(_) => Ok(()),
+                Err(e) => Err(Error::EmitterError(e.to_string())),
+            };
+        }
+
+        Ok(())
+    }
+
+    /// Attempt to send all events currently in the event store
+    fn flush(&mut self) -> Result<(), Error> {
+        // Get a batch of all events currently in the event store
+        let batch = {
+            let mut store = match self.event_store.lock() {
+                Ok(store) => store,
+                Err(e) => return Err(Error::EmitterError(e.to_string())),
+            };
+            let queue_length = store.len();
```

Comment:
Nit: I think this could be made more atomic if instead of calling `len()` and then `batch_of(len)`, we had a single call to return the complete batch (or a list of batches) to send. Also, not very familiar with Rust locking, but shouldn't this be inside the `lock() { }` block?

Reply:
This does still hold the lock until `store` is dropped, but I have simplified this anyway to avoid the unnecessary extra block.